native/src/widget/table.rs

```diff
@@ -0,0 +1,188 @@
+    pub fn push(mut self, row: Vec<Element<'a, Message, Renderer>>) -> Self {
+        self.rows.push(row);
+        self
+    }
+}
+
+impl<'a, Message, Renderer> Widget<Message, Renderer>
+    for Table<'a, Message, Renderer>
+where
+    Renderer: self::Renderer,
+{
+    fn width(&self) -> Length {
+        Length::Shrink
+    }
+
+    fn height(&self) -> Length {
+        Length::Shrink
+    }
+
+    fn layout(
+        &self,
+        renderer: &Renderer,
+        limits: &layout::Limits,
+    ) -> layout::Node {
+        // store calculated layout sizes
+        let mut layouts = Vec::with_capacity(self.rows.len());
+        // store width of each column
+        let mut column_widths = Vec::<f32>::new();
+
+        // find out how wide a column is by finding the widest cell in it
+        for row in &self.rows {
+            let mut layouts_row = Vec::with_capacity(row.len());
+
+            for (index, cell) in row.iter().enumerate() {
+                let layout = cell.layout(renderer, limits).size();
+                layouts_row.push(layout);
+
+                if let Some(column_width) = column_widths.get_mut(index) {
+                    *column_width = column_width.max(layout.width);
+                } else {
+                    column_widths.insert(index, layout.width);
+                }
+            }
+
+            layouts.push(layouts_row);
+        }
+
+        // list of x alignments for every column
+        let column_aligns: Vec<_> = iter::once(&0.)
+            .chain(&column_widths)
+            .scan(0., |state, width| {
+                *state += width;
+                Some(*state)
+            })
+            .collect();
+
+        let mut nodes_table = Vec::with_capacity(self.rows.len());
+        let table_width = column_widths.into_iter().sum();
+        let mut table_height = 0.;
+
+        for row in layouts {
+            let mut nodes_row = Vec::with_capacity(row.len());
+            let mut row_height: f32 = 0.;
+
+            for (size, column_align) in row.into_iter().zip(&column_aligns) {
+                let node = Node::new(size, Size::new(*column_align, 0.));
+                nodes_row.push(node);
+                row_height = row_height.max(size.height);
+            }
+
+            nodes_table.push(Node::with_children(
+                Size::new(table_width, row_height),
+                Size::new(0., table_height),
+                nodes_row,
+            ));
+
+            table_height += row_height;
+        }
+
+        Node::with_children(
+            Size::new(table_width, table_height),
+            Size::ZERO,
+            nodes_table,
+        )
+    }
+
+    fn draw(
+        &self,
+        renderer: &mut Renderer,
+        defaults: &Renderer::Defaults,
+        layout: Layout<'_>,
+        cursor_position: Point,
+    ) -> Renderer::Output {
+        renderer.draw(defaults, layout, cursor_position, &self.rows)
+    }
+
+    fn hash_layout(&self, state: &mut Hasher) {
```

Comment:
I saw in other implementations a line like this:
```rust
0.hash(state);
```
Not sure what it is supposed to do.

Reply:
That is a leftover that we should fix eventually.

The idea is that we have to include a unique identifier for this widget in the resulting hash. We can achieve this with:

```rust
std::any::TypeId::of::<Grid<'static, ()>>().hash(state);
```

Reply:
Done.